## 读取数据

简介：  
Contains orders from an E-commerce site (UK Retailer), with various products and orders. Try starting with EDA and applying different functions to know more about the data. The dataset is tricky as it contains all the problems that an actual site may have.

Some hints to know what problems the dataset has:

Contains duplicates!! Check for it and remove them
Contains missing values
Has negative unit price, try to guess what could be the reason
Has negative quantity, try to guess what could be the reason
Apply a prescriptive analysis approach to data and find ways to increase sales. This is an interesting task that builds up skills like business Intelligence.  

分析目标：  
根据市场销售数据，挖掘畅销产品，以便制定更有效的市场策略来提成营收。  

In [1]:
import pandas as pd

In [11]:
original_data = pd.read_csv("data.csv", encoding="Shift_JIS")

In [12]:
original_data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

## 评估数据

### 评估数据整齐度

In [13]:
original_data.sample(10)

InvoiceNo StockCode                         Description  Quantity  \
336176    566307     23083    SET 6 PAPER TABLE LANTERN STARS         12   
211244    555350     20728                 LUNCH BAG CARS BLUE        10   
151183    549546     22938            CUPCAKE LACE PAPER SET 6         4   
427260    573380     23128         FELTCRAFT BOY JEAN-PAUL KIT         1   
297762    562955     21218              RED SPOTTY BISCUIT TIN        16   
538860    581419     22441   GROW YOUR OWN BASIL IN ENAMEL MUG         8   
239674    558066     22329  ROUND CONTAINER SET OF 5 RETROSPOT         1   
148285    549222     22376   AIRLINE BAG VINTAGE JET SET WHITE         4   
308561    564037     22383              LUNCH BAG SUKI DESIGN         10   
72749     542258     22720   SET OF 3 CAKE TINS PANTRY DESIGN          2   

             InvoiceDate  UnitPrice  CustomerID         Country  
336176    9/12/2011 9:37       3.75     14307.0  United Kingdom  
211244    6/2/2011 12:08       1.65     17652.0  United Kingdom  
151183   4/10/2011 11:24       1.95     16889.0  United Kingdom  
427260  10/30/2011 14:04       4.95     14572.0  United Kingdom  
297762   8/11/2011 10:14       8.29         NaN  United Kingdom  
538860   12/8/2011 15:03       2.10     15484.0  United Kingdom  
239674   6/24/2011 15:45       3.29         NaN  United Kingdom  
148285    4/7/2011 10:43       4.25     12347.0         Iceland  
308561   8/22/2011 12:26       1.65     14800.0  United Kingdom  
72749    1/26/2011 17:08       4.95     17372.0  United Kingdom

从抽样的10行数据来看。符合“每列是一个变量，每行是一个观察值，每个单元格是一个值”，不存在结构性问题。

### 评估数据的干净程度

In [14]:
original_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


从输出结果看，数据共有541909条观察值，而`descriotion`、`CustomerID`变量存在缺失值。  
此外，`InvoiceDate`的数据类型应为日期，`CustomerID`的数据类型应为字符串，应当进行数据格式转换。

#### 评估缺失数据

In [15]:
original_data[original_data["Description"].isnull()]

InvoiceNo StockCode Description  Quantity      InvoiceDate  UnitPrice  \
622       536414     22139         NaN        56  12/1/2010 11:52        0.0   
1970      536545     21134         NaN         1  12/1/2010 14:32        0.0   
1971      536546     22145         NaN         1  12/1/2010 14:33        0.0   
1972      536547     37509         NaN         1  12/1/2010 14:33        0.0   
1987      536549    85226A         NaN         1  12/1/2010 14:34        0.0   
...          ...       ...         ...       ...              ...        ...   
535322    581199     84581         NaN        -2  12/7/2011 18:26        0.0   
535326    581203     23406         NaN        15  12/7/2011 18:31        0.0   
535332    581209     21620         NaN         6  12/7/2011 18:35        0.0   
536981    581234     72817         NaN        27  12/8/2011 10:33        0.0   
538554    581408     85175         NaN        20  12/8/2011 14:06        0.0   

        CustomerID         Country  
622            NaN  United Kingdom  
1970           NaN  United Kingdom  
1971           NaN  United Kingdom  
1972           NaN  United Kingdom  
1987           NaN  United Kingdom  
...            ...             ...  
535322         NaN  United Kingdom  
535326         NaN  United Kingdom  
535332         NaN  United Kingdom  
536981         NaN  United Kingdom  
538554         NaN  United Kingdom  

[1454 rows x 8 columns]

有1454条交易数据缺失Descriotion的交易数据，Unitprice都为0.为了验证才行，我们增加筛选条件，看是否存在Descriotion变量缺失且Unitprice不为0的数据。

In [19]:
original_data[(original_data["Description"].isnull()) & (original_data["UnitPrice"] != 0)]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

筛选出来的结果为0条，说明缺失Description值的数据同时也不具备有效的UnitPrice值。

Description表示产品名称，UnitPrice表示产品单价，都是进行后续商品交易分析的重要变量。如果他们同时缺失/无效，我们认为数据无法提供有效含义，因此这些数据可以被删除。

In [20]:
original_data[original_data["CustomerID"].isnull()]

InvoiceNo StockCode                      Description  Quantity  \
622       536414     22139                              NaN        56   
1443      536544     21773  DECORATIVE ROSE BATHROOM BOTTLE         1   
1444      536544     21774  DECORATIVE CATS BATHROOM BOTTLE         2   
1445      536544     21786               POLKADOT RAIN HAT          4   
1446      536544     21787            RAIN PONCHO RETROSPOT         2   
...          ...       ...                              ...       ...   
541536    581498    85099B          JUMBO BAG RED RETROSPOT         5   
541537    581498    85099C   JUMBO  BAG BAROQUE BLACK WHITE         4   
541538    581498     85150    LADIES & GENTLEMEN METAL SIGN         1   
541539    581498     85174                S/4 CACTI CANDLES         1   
541540    581498       DOT                   DOTCOM POSTAGE         1   

            InvoiceDate  UnitPrice  CustomerID         Country  
622     12/1/2010 11:52       0.00         NaN  United Kingdom  
1443    12/1/2010 14:32       2.51         NaN  United Kingdom  
1444    12/1/2010 14:32       2.51         NaN  United Kingdom  
1445    12/1/2010 14:32       0.85         NaN  United Kingdom  
1446    12/1/2010 14:32       1.66         NaN  United Kingdom  
...                 ...        ...         ...             ...  
541536  12/9/2011 10:26       4.13         NaN  United Kingdom  
541537  12/9/2011 10:26       4.13         NaN  United Kingdom  
541538  12/9/2011 10:26       4.96         NaN  United Kingdom  
541539  12/9/2011 10:26      10.79         NaN  United Kingdom  
541540  12/9/2011 10:26    1714.17         NaN  United Kingdom  

[135080 rows x 8 columns]

CustomerID表示客户编号，不是分析畅销产品的必要变量。并且从输出结果看，有些CustomerID缺失的交易数据仍然有效，因此保留次变量为空的观察值。

### 评估重复数据

根据数据变量的含义来看，虽然InvoiceNo、StockCode和CustomerID都是唯一标识符，但一次交易可能包含多件商品，因此InvoiceNo可以存在重复，不同交易可以包含同件商品，因此StockCode可以存在重复，顾客可以进行多次交易或下单多个商品，因此CustomerID也可以存在重复。  
那么针对次数据集，我们无需评估重复数据。

### 评估不一致数据

不一致数据可能存在于country中，我们要检查是否存在多个不同值指代同一国家的情况。

In [22]:
original_data["Country"].value_counts()

Country
United Kingdom          495478
Germany                   9495
France                    8557
EIRE                      8196
Spain                     2533
Netherlands               2371
Belgium                   2069
Switzerland               2002
Portugal                  1519
Australia                 1259
Norway                    1086
Italy                      803
Channel Islands            758
Finland                    695
Cyprus                     622
Sweden                     462
Unspecified                446
Austria                    401
Denmark                    389
Japan                      358
Poland                     341
Israel                     297
USA                        291
Hong Kong                  288
Singapore                  229
Iceland                    182
Canada                     151
Greece                     146
Malta                      127
United Arab Emirates        68
European Community          61
RSA                         58


从Country变量值来看，“USA”、“United States”均表示美国，“United Kingdom”、”UK“、”U.K.“均在表示英国，因此应该在清洗步骤对这些值进行统一，只保留一个指代值。

### 评价无效或错误数据

可以通过DataFrame的“describe”方法，对数值统计信息进行快速了解。

In [23]:
original_data.describe()

Quantity      UnitPrice     CustomerID
count  541909.000000  541909.000000  406829.000000
mean        9.552250       4.611114   15287.690570
std       218.081158      96.759853    1713.600303
min    -80995.000000  -11062.060000   12346.000000
25%         1.000000       1.250000   13953.000000
50%         3.000000       2.080000   15152.000000
75%        10.000000       4.130000   16791.000000
max     80995.000000   38970.000000   18287.000000

从输出结果来看，Quantity和UnitPrice存在附属，会对后续数值分析产生影响。

因此，我们先筛选出Quantity数值为负数的观察值，进一步评估其含义。

In [24]:
original_data[original_data["Quantity"] < 0]

InvoiceNo StockCode                       Description  Quantity  \
141      C536379         D                          Discount        -1   
154      C536383    35004C   SET OF 3 COLOURED  FLYING DUCKS        -1   
235      C536391     22556    PLASTERS IN TIN CIRCUS PARADE        -12   
236      C536391     21984  PACK OF 12 PINK PAISLEY TISSUES        -24   
237      C536391     21983  PACK OF 12 BLUE PAISLEY TISSUES        -24   
...          ...       ...                               ...       ...   
540449   C581490     23144   ZINC T-LIGHT HOLDER STARS SMALL       -11   
541541   C581499         M                            Manual        -1   
541715   C581568     21258        VICTORIAN SEWING BOX LARGE        -5   
541716   C581569     84978  HANGING HEART JAR T-LIGHT HOLDER        -1   
541717   C581569     20979     36 PENCILS TUBE RED RETROSPOT        -5   

            InvoiceDate  UnitPrice  CustomerID         Country  
141      12/1/2010 9:41      27.50     14527.0  United Kingdom  
154      12/1/2010 9:49       4.65     15311.0  United Kingdom  
235     12/1/2010 10:24       1.65     17548.0  United Kingdom  
236     12/1/2010 10:24       0.29     17548.0  United Kingdom  
237     12/1/2010 10:24       0.29     17548.0  United Kingdom  
...                 ...        ...         ...             ...  
540449   12/9/2011 9:57       0.83     14397.0  United Kingdom  
541541  12/9/2011 10:28     224.69     15498.0  United Kingdom  
541715  12/9/2011 11:57      10.95     15311.0  United Kingdom  
541716  12/9/2011 11:58       1.25     17315.0  United Kingdom  
541717  12/9/2011 11:58       1.25     17315.0  United Kingdom  

[10624 rows x 8 columns]

从筛选结果来看，Quantity变量为负数的观察值，InvoiceNo似乎均以C开头，表示订单被取消。

但此猜测不一定准确，为了验证猜想，我们增加筛选条件，看是否存在Quantity变量为负数且InvoiceNo不以C开头的观察值。

In [25]:
original_data[(original_data["Quantity"] < 0) & (original_data["InvoiceNo"].str[0] != "C")]

InvoiceNo StockCode Description  Quantity      InvoiceDate  UnitPrice  \
2406      536589     21777         NaN       -10  12/1/2010 16:50        0.0   
4347      536764    84952C         NaN       -38  12/2/2010 14:42        0.0   
7188      536996     22712         NaN       -20  12/3/2010 15:30        0.0   
7189      536997     22028         NaN       -20  12/3/2010 15:30        0.0   
7190      536998     85067         NaN        -6  12/3/2010 15:30        0.0   
...          ...       ...         ...       ...              ...        ...   
535333    581210     23395       check       -26  12/7/2011 18:36        0.0   
535335    581212     22578        lost     -1050  12/7/2011 18:38        0.0   
535336    581213     22576       check       -30  12/7/2011 18:38        0.0   
536908    581226     23090     missing      -338   12/8/2011 9:56        0.0   
538919    581422     23169     smashed      -235  12/8/2011 15:24        0.0   

        CustomerID         Country  
2406           NaN  United Kingdom  
4347           NaN  United Kingdom  
7188           NaN  United Kingdom  
7189           NaN  United Kingdom  
7190           NaN  United Kingdom  
...            ...             ...  
535333         NaN  United Kingdom  
535335         NaN  United Kingdom  
535336         NaN  United Kingdom  
536908         NaN  United Kingdom  
538919         NaN  United Kingdom  

[1336 rows x 8 columns]

以上猜想被证实错误，因为还存在Quantity变量为负数且InvoiceNo不以C开头的观察值。

但根据以上输出结果，这些筛选出来的观察值的UnitPrice均为0，因此增加UnitPrice的条件进行验证。

In [27]:
original_data[(original_data["Quantity"] < 0) & (original_data["InvoiceNo"].str[0] != "C") & (original_data["UnitPrice"] != 0)]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

根据输出结果，猜想得到验证，Quantity变量为负数，观察值满足以下条件之一：  
- InvoiceNo以C开头，表示订单被取消
- UnitPrice为0，说明单价为0英镑

这些交易数据均不是有效成交数据，不贡献销售，不在后续分析范围内，因此我们将在数据清理步骤，将Quantity为负数的观察值删除。

In [28]:
original_data[original_data["UnitPrice"] < 0]

InvoiceNo StockCode      Description  Quantity      InvoiceDate  \
299983   A563186         B  Adjust bad debt         1  8/12/2011 14:51   
299984   A563187         B  Adjust bad debt         1  8/12/2011 14:52   

        UnitPrice  CustomerID         Country  
299983  -11062.06         NaN  United Kingdom  
299984  -11062.06         NaN  United Kingdom

从输出结果来看，UnitPrice为负数的观察值都是坏账调整，不属于实际商品交易数据，因此也在数据清理步骤中将其删除。

## 清洗数据

根据前面的评估部分得到的结论，我们需要进行的数据清理包括：  
- 把InboiceDate变量的数据类型转换为日期时间
- 把CustomerID变量的数据类型转换为字符串
- 把Description变量缺失的观察值删除
- 把Country变量值“USA”替换为“United States”
- 把Country变量值“UK”、“U.K.”替换为“United Kingdom”
- 把Quantity变量值为负数的观察值删除
- 把“UnitPrice”变量值为负数的观察值删除

为了区分开经过清理的数据和原始数据，我们创建新的扮靓cleaned_data，让它为original_data复制出的副本。我们之后的清理步骤都将运用在cleaned_data上。 

In [32]:
cleaned_data = original_data.copy()

将InboiceDate变量的数据类型转换为日期时间

In [34]:
cleaned_data["InvoiceDate"] = pd.to_datetime(cleaned_data["InvoiceDate"])
cleaned_data["InvoiceDate"]

0        2010-12-01 08:26:00
1        2010-12-01 08:26:00
2        2010-12-01 08:26:00
3        2010-12-01 08:26:00
4        2010-12-01 08:26:00
                 ...        
541904   2011-12-09 12:50:00
541905   2011-12-09 12:50:00
541906   2011-12-09 12:50:00
541907   2011-12-09 12:50:00
541908   2011-12-09 12:50:00
Name: InvoiceDate, Length: 541909, dtype: datetime64[ns]

把CustomerID变量的数据类型转换为字符串

In [35]:
cleaned_data["CustomerID"] = cleaned_data["CustomerID"].astype(str)
cleaned_data["CustomerID"]

0         17850.0
1         17850.0
2         17850.0
3         17850.0
4         17850.0
           ...   
541904    12680.0
541905    12680.0
541906    12680.0
541907    12680.0
541908    12680.0
Name: CustomerID, Length: 541909, dtype: object

把CustomerID变量值结尾的“.0”删除：

In [38]:
cleaned_data["CustomerID"] = cleaned_data["CustomerID"].str.slice(0, -2)

把Description变量缺失的观察值删除

In [39]:
cleaned_data = cleaned_data.dropna(subset = ["Description"])

In [40]:
cleaned_data["Description"].isnull().sum()

0

把Country变量值“USA”替换为“United States”

In [51]:
cleaned_data.loc[:, "Country"] = cleaned_data["Country"].replace({"USA": "United States"})

In [42]:
len(cleaned_data[cleaned_data["Country"] == "USA"])

0

把Country变量值“UK”、“U.K.”替换为“United Kingdom”

In [44]:
cleaned_data.loc[:, "Country"] = cleaned_data["Country"].replace({"UK": "United Kingdom", "U.K.": "United Kingdom"})

In [45]:
print(len(cleaned_data[cleaned_data["Country"] == "UK"]))
print(len(cleaned_data[cleaned_data["Country"] == "U.K."]))

0
0


把Quantity变量值为负数的观察值删除

In [46]:
cleaned_data = cleaned_data[cleaned_data["Quantity"] >= 0]

In [47]:
len(cleaned_data[cleaned_data["Quantity"] < 0])

0

把“UnitPrice”变量值为负数的观察值删除

In [48]:
cleaned_data = cleaned_data[cleaned_data["UnitPrice"] >= 0]

In [49]:
len(cleaned_data[cleaned_data["Quantity"] < 0])

0

## 保存清理后的数据

In [50]:
cleaned_data.to_csv("data_cleaned.csv", index = False)